### PR TITLE
Python tix module

### PR DESCRIPTION
**Pull request by symphorien**
###### Motivation for this change
 Being able to use the [tkinter.tix](https://docs.python.org/3.5/library/tkinter.tix.html) python module (which is in the standard library).
This is done in three parts:
* adding tk private headers to the dev output of the derivation (tcl private headers are already there) which are needed to build tix
* adding a tix package
* wrapping python to set the correct [TIX_LIBRARY](https://docs.python.org/3.5/library/tkinter.tix.html#using-tix) environment variable

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
Regarding testing : I rebuilt my whole system with those versions of cpython (my laptop has not enough ram to build pypy) and it has been working fine since then. I could not test extensively all the packages impacted by those changes. And of course I checked that the tkinter.tix module works.

### CODE REVIEW

File: pkgs/development/interpreters/python/cpython/2.7/default.nix [hunk 1]
```diff
@@ -121,7 +122,7 @@ let
 
   buildInputs =
     optional (stdenv ? cc && stdenv.cc.libc != null) stdenv.cc.libc ++
-    [ bzip2 openssl zlib ]
+    [ bzip2 openssl zlib makeWrapper]
     ++ optionals stdenv.isCygwin [ expat libffi ]
     ++ [ db gdbm ncurses sqlite readline ]
     ++ optionals x11Support [ tcl tk xlibsWrapper libX11 ]
```
Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix]: ` ++ optionals x11Support [ tcl tk libX11 xproto tix ]`

Then you should be able to get rid of the wrapper. At least, I made the change and it seems to work.

File: pkgs/development/libraries/tix/default.nix [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  '';
+
+  meta = with stdenv.lib; {
+    description = "A widget library for Tcl/Tk";
+    homepage    = http://tix.sourceforge.net/;
+    platforms   = platforms.all;
+    license     = with licenses; [
+      bsd2 # tix
+      gpl2 # patches from portage
+    ];
```
Comment (FRidh) [inline pkgs/development/libraries/tix/default.nix]: would you like to maintain this package?

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -9493,6 +9493,14 @@ in
 
   tivodecode = callPackage ../applications/video/tivodecode { };
 
+  tix = tix-8_6;
+
+  tix-8_6 = callPackage ../development/libraries/tix { };
+  tix-8_5 = callPackage ../development/libraries/tix {
```
Comment (FRidh) [inline pkgs/top-level/all-packages.nix]: We only use tix-8_6, so we don't need tix-8_5 and can just use 
```
tix = callPackage ...
```

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by FRidh** [comment]
I don't think this will be a mass rebuild (tk isn't used much) but even so please rebase against staging.

File: pkgs/development/interpreters/python/cpython/2.7/default.nix (hunk 1 above)
Comment (symphorien) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix]: Not sure to understand well : Do you want me to put makeWrapper in `[ tcl tk libX11 xproto tix ]` ?

File: pkgs/development/libraries/tix/default.nix (hunk 2 above)
Comment (symphorien) [inline pkgs/development/libraries/tix/default.nix]: No.

File: pkgs/development/interpreters/python/cpython/2.7/default.nix (hunk 1 above)
Comment (FRidh) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix]: no, put tix there like I wrote.

Comment (symphorien) [inline pkgs/development/interpreters/python/cpython/2.7/default.nix]: I tried this http://xelpaste.net/2J0wKk but it doesn't work for me : 
```
    self.tk.eval('package require Tix')
_tkinter.TclError: can't find package Tix
```

**Comment by FRidh** [comment]
Indeed, you are right. I checked the source code and it always tries to find the library using an environment variable.

We shouldn't use a wrapper if we can do so without. I think we can patch the code to point to the Tix library.

The following (Python 3.5) doesn't seem to be sufficient yet
```
  postPatch = optionalString x11Support ''
    substituteInPlace "Lib/tkinter/tix.py" --replace "os.environ.get('TIX_LIBRARY')" "'${tix}/lib'"
  '';
```
because
```
>>> import tkinter.tix
>>> root = tkinter.tix.Tk()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/p6m8pxyw6njqfk82bdgj08w8pdrhq17h-python3-3.5.2/lib/python3.5/tkinter/tix.py", line 221, in __init__
    self.tk.eval('package require Tix')
_tkinter.TclError: Can't find a usable Init.tcl in the following directories: 
    /nix/store/aqfcswnix9ayq1jwrh931gjimnwxc7mq-tcl-8.6.6/lib/tcl8.6/Tix8.4.3 /nix/store/aqfcswnix9ayq1jwrh931gjimnwxc7mq-tcl-8.6.6/lib/Tix8.4.3 /nix/store/p6m8pxyw6njqfk82bdgj08w8pdrhq17h-python3-3.5.2/lib/Tix8.4.3 /nix/store/8djjhdby3grc0zq90bl5qg7ckzvgrmfd-tk-8.6.6/lib/tk8.6/Tix8.4.3 /nix/store/8djjhdby3grc0zq90bl5qg7ckzvgrmfd-tk-8.6.6/lib/tk8.6/ttk/Tix8.4.3 /nix/store/p6m8pxyw6njqfk82bdgj08w8pdrhq17h-python3-3.5.2/bin/Tix8.4.3 /nix/store/95y3xa78f7ksdy9hwks01axhw738km85-tix-8.4.3/Tix8.4.3 /nix/store/lib/Tix8.4.3 /nix/store/p6m8pxyw6njqfk82bdgj08w8pdrhq17h-python3-3.5.2/library
```

**Comment by symphorien** [comment]
The patch you proposed should have worked, but there was a mistake in the way I packaged tix for it to work. I force pushed something which is close to what you proposed.

File: pkgs/development/interpreters/python/pypy/2.7/default.nix [hunk 4]
```diff
@@ -113,6 +113,7 @@ let
        # will think the wrapper is the original. As long as the wrapper has
        # the same path as the original, this is OK.
        wrapProgram "$out/pypy-c/pypy-c" \
+         --set TIX_LIBRARY ${tix}/lib \
```
Comment (FRidh) [inline pkgs/development/interpreters/python/pypy/2.7/default.nix]: grepping the PyPy source code gave me the following file that needs to be patched
```
lib-python/2.7/lib-tk/Tix.py:        tixlib = os.environ.get('TIX_LIBRARY')
```

**Review by FRidh — CHANGES_REQUESTED** [review]
Great. I see you want to keep support for `TIX_LIBRARY`? That seems reasonable.

**Comment by FRidh** [comment]
There is one more challenge though. The default Python `python` won't have the reference to Tix, which is what we want. `pythonFull` will have the reference, that is fine, but `pythonFull` is deprecated in favor of using `python.withPackages(ps: [ps.tkinter])`. And the latter won't work yet
```
$ nix-shell -p 'python.withPackages(ps: [ps.tkinter])' --run "python -c 'import Tix; Tix.Tk()'" -I nixpkgs=.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/nix/store/i1m970i9mbrk0aqr5z5rf0rfx8jihmy4-python-2.7.13-env/lib/python2.7/lib-tk/Tix.py", line 222, in __init__
    self.tk.eval('package require Tix')
_tkinter.TclError: can't find package Tix
```
because it uses the interpreter from `python` and the `_tkinter` module from `pythonFull`. (Note that I just saw it also links to the Python interpreter from `pythonFull` which is a bit of a problem.)

**Comment by symphorien** [comment]
I forgot pypy, that should be fixed.
Regarding pythonFull, do you mean that the x11support value is to be eventually removed ?

**Comment by FRidh** [comment]
> Regarding pythonFull, do you mean that the x11support value is to be eventually removed ?

The parameter will stay; just the attibute `pkgs.pythonFull` is deprecated. The parameter is used to actually be able to build `pythonPackages.tkinter`.

**Review by FRidh — APPROVED** [review]
> (Note that I just saw it also links to the Python interpreter from pythonFull which is a bit of a problem.)

This has been fixed now in https://github.com/NixOS/nixpkgs/commit/e276f590200e03b6e733c4e680ba2d835e37cd56

@symphorien This PR looks good to me as it is. Unfortunately `tix` won't work yet with `python.withPackages(ps: [ps.tkinter])`. Even so I think this can go in now.

**Pull request merged**